include/dlaf/permutations/general/impl.h

```diff
@@ -0,0 +1,201 @@
+
+      dlaf::tile::lacpy<T>(region, i_subtile_in, tile_in, i_subtile_out, tile_out);
+    }
+  });
+}
+
+#if defined(DLAF_WITH_CUDA)
+template <class T, Coord coord>
+void applyPermutations(GlobalElementIndex out_begin, GlobalElementSize sz, SizeType in_offset,
+                       const matrix::Distribution& distr, const SizeType* perm_arr,
+                       const std::vector<matrix::Tile<T, Device::GPU>>& in_tiles,
+                       const std::vector<matrix::Tile<T, Device::GPU>>& out_tiles, cudaStream_t stream) {
+  constexpr Coord ocrd = orthogonal(coord);
+  std::vector<SizeType> splits =
+      util::interleaveSplits(sz.get<ocrd>(), distr.blockSize().get<ocrd>(),
+                             distr.distanceToAdjacentTile<ocrd>(in_offset),
+                             distr.distanceToAdjacentTile<ocrd>(out_begin.get<ocrd>()));
+
+  for (SizeType i_perm = 0; i_perm < sz.get<coord>(); ++i_perm) {
+    for (std::size_t i_split = 0; i_split < splits.size() - 1; ++i_split) {
+      SizeType split = splits[i_split];
+
+      GlobalElementIndex i_split_gl_in(split + in_offset, perm_arr[i_perm]);
+      GlobalElementIndex i_split_gl_out(split + out_begin.get<ocrd>(), out_begin.get<coord>() + i_perm);
+      TileElementSize region(splits[i_split + 1] - split, 1);
+      if constexpr (coord == Coord::Row) {
+        region.transpose();
+        i_split_gl_in.transpose();
+        i_split_gl_out.transpose();
+      }
+
+      TileElementIndex i_subtile_in = distr.tileElementIndex(i_split_gl_in);
+      auto& tile_in = in_tiles[to_sizet(distr.globalTileLinearIndex(i_split_gl_in))];
+      TileElementIndex i_subtile_out = distr.tileElementIndex(i_split_gl_out);
+      auto& tile_out = out_tiles[to_sizet(distr.globalTileLinearIndex(i_split_gl_out))];
+
+      dlaf::gpulapack::lacpy<T>(blas::Uplo::General, region.rows(), region.cols(),
+                                tile_in.ptr(i_subtile_in), tile_in.ld(), tile_out.ptr(i_subtile_out),
+                                tile_out.ld(), stream);
+    }
+  };
+}
+#endif
+
+template <Backend B, Device D, class T, Coord C>
+void Permutations<B, D, T, C>::call(SizeType i_begin, SizeType i_end, Matrix<const SizeType, D>& perms,
+                                    Matrix<T, D>& mat_in, Matrix<T, D>& mat_out) {
+  namespace ut = matrix::util;
+  namespace ex = pika::execution::experimental;
+
+  const matrix::Distribution& distr = mat_in.distribution();
+  TileElementSize sz_last_tile = distr.tileSize(GlobalTileIndex(i_end, i_end));
+  SizeType m = distr.globalTileElementDistance<Coord::Row>(i_begin, i_end) + sz_last_tile.rows();
+  SizeType n = distr.globalTileElementDistance<Coord::Col>(i_begin, i_end) + sz_last_tile.cols();
+  matrix::Distribution subm_distr(LocalElementSize(m, n), distr.blockSize());
+  SizeType ntiles = i_end - i_begin + 1;
+
+  // Note: there is an issue compiling with dlaf::internal::transform()
```

Comment:
For this I'd try replacing https://github.com/eth-cscs/DLA-Future/blob/60c620713feaaf6590e823947f412b5254609685/include/dlaf/sender/transform.h#L80 with `pika::unwrapping_n<1>(...)` (in the hope that we don't have other places that depend on having nested futures unwrapped automatically).

Reply:
I tried this but it still resulted in a compile-time error.

Reply:
Hmm, ok. I will try to see what's causing problems there.

Reply:
This is not a blocker. I'll look into this as soon as possible.

Reply:
So if I understand the `unwrapping` implementation correctly, the "depth" actually refers to the depth of unwrapping specifically futures which means that the `shared_future` in `std::vector<pika::shared_future<...>>` is considered to be at depth 1 and will be unwrapped.

As a short term solution I would suggest something like https://github.com/msimberg/DLA-Future/commit/e7297d3ede222df11f1138ba64dba1979e99e1e5, which simply adds an option to tell `transform` not to unwrap. That commit only adds it to the basic `transform` overload and nowhere else.